### DISCUSSION TITLE
Consolidating AppStoreConnect credentials code

### DISCUSSION THREAD

**Discussion by lacostej** [fastlane/fastlane#18145] [Ideas]
We've activated 2FA and are now setting up AppStoreConnect API accross the board.

Some notes and questions related to my findings

* there seems to be quite a few questions related to AppStore Connect right now.

* The current recommendation is to use the `app_store_connect_api_key` action to set the `APP_STORE_CONNECT_API_KEY` lane context variable.

* there's quite a bit of duplicated code in fastlane doing things like

```
def api_token
  @api_token ||= Spaceship::ConnectAPI::Token.create(Precheck.config[:api_key]) if Precheck.config[:api_key]
  @api_token ||= Spaceship::ConnectAPI::Token.from_json_file(Precheck.config[:api_key_path]) if Precheck.config[:api_key_path]
  return @api_token
end
```

* each tool seems to have defined its own environment variable to pass the key path
```
$ git grep API_KEY_PATH
cert/lib/cert/options.rb:                                     env_name: "DELIVER_API_KEY_PATH",
deliver/lib/deliver/options.rb:                                     env_name: "DELIVER_API_KEY_PATH",
fastlane/lib/fastlane/actions/app_store_build_number.rb:                                       env_name: "APPSTORE_BUILD_NUMBER_API_KEY_PATH",
fastlane/lib/fastlane/actions/latest_testflight_build_number.rb:                                       env_name: "APPSTORE_BUILD_NUMBER_API_KEY_PATH",
fastlane/lib/fastlane/actions/register_device.rb:                                       env_name: "FL_REGISTER_DEVICE_API_KEY_PATH",
fastlane/lib/fastlane/actions/register_devices.rb:                                       env_name: "FL_REGISTER_DEVICES_API_KEY_PATH",
fastlane/lib/fastlane/actions/set_changelog.rb:                                     env_name: "FL_SET_CHANGELOG_API_KEY_PATH",
match/lib/match/options.rb:                                     env_name: "SIGH_API_KEY_PATH",
pilot/lib/pilot/options.rb:                                     env_name: "PILOT_API_KEY_PATH",
precheck/lib/precheck/options.rb:                                     env_name: "PRECHECK_API_KEY_PATH",
sigh/lib/sigh/options.rb:                                     env_name: "SIGH_API_KEY_PATH",
```
* no guidelines exposes `Spaceship::ConnectAPI::Token.from_json_file()`  which to me would be a good candidate for use as it proposes a simpler API compared to  `app_store_connect_api_key` and allows me to put the whole credentials file into a vault (e.g. cryptex). A variant that takes the content might be useful as well for people who don't want to store [on disk](https://github.com/fastlane/fastlane/discussions/18115)

* there's no state in `Spaceship::ConnectAPI.token`. If I set it in my Fastfile, fastlane doesn't detect it is already setup. I experimented with [this patch](https://github.com/lacostej/fastlane/commits/feature/reuse_api_connect_token). It allows me to set the credentials like that.
```
    if !ENV['FASTLANE_ITC_API_KEY_PATH'].nil?
      fastlane_json_credentials = ENV['FASTLANE_ITC_API_KEY_PATH']
      UI.message "loading ITC API_KEY_PATH from #{fastlane_json_credentials}"
      require 'spaceship'
      Spaceship::ConnectAPI.token = Spaceship::ConnectAPI::Token.from_json_file(fastlane_json_credentials)
    end
```
Note on this, another way to achieve the same thing without patching fastlane is to do the following. Note that `app_store_connect_api_key` (that sets a hash representing the token to be converted to a token object later) and `from_json_file` (that creates a token object right away) have a slight incompatibly in their expectations.

```
      api_key = JSON.parse(File.read(fastlane_json_credentials), { symbolize_names: true })
      api_key[:key_content] = api_key.delete(:key) # rename key to match app_store_connect_api_key API
      app_store_connect_api_key(api_key)
```

* some people have also raised the issue of the `APP_STORE_CONNECT_API_KEY` [leaking into logs](https://github.com/fastlane/fastlane/discussions/18125). I haven't looked at this yet.

* people would like to use a place to [share those ITC connect keys ](https://github.com/fastlane/fastlane/discussions/18093). I think we will use cryptex for now as we already use it for our Android/Windows/... infra.

From all of this, it looks like there could be some consolidation. Some ideas for a first round of cleanups:
* provide a unified ENV variable setup for passing the credentials to all fastlane ecosystem
* provide a way to set up the key from an action using a single parameter, the file or the file content
* consolidate the duplicated code, and support programmatic setup of the `Spaceship::ConnectAPI.token`
* ensure there's no leak of any of the keys in the console.

Finally I am a bit worried that several features are not supported by the official API. There doesn't seem to be a work-around apart from the hackish session management if Apple enforces 2FA fully. Some tooling to share this session might be useful.

Thoughts anyone? (CC @joshdholtz)

**Comment by joshdholtz** [comment]
@lacostej There TOTALLY needs to be consolidation 😊 I should really get on that this week.

One of the reasons I did not do this eventually was because the App Store Connect API wasn't not supported evenly and I wanted users to have the flexibility to use it explicitly in the tools they wanted to. In hindsight, I should have had a tool specific environment variable and a global one 🤦‍♂️

I'm going to start a change for this right now!

**Reply by joshdholtz** [reply]
@lacostej The leak you mentioned (from https://github.com/fastlane/fastlane/discussions/18125) is no longer an issue 😇 That was fixed in `2.161.0` and that discussion has been solved/answered now 💪
